src/client/opamAction.ml

```diff
@@ -366,15 +366,27 @@ let prepare_package_source st nv dir =
   in
   let check_extra_files =
     try
-      List.iter (fun (src,base,hash) ->
-          if not (OpamHash.check_file (OpamFilename.to_string src) hash) then
-            failwith
-              (Printf.sprintf "Bad hash for %s" (OpamFilename.to_string src))
+      let extra_files_dir =
+        OpamPath.Switch.extra_files_dir st.switch_global.root st.switch
+      in
+      List.iter (fun (base, hash) ->
+          let dst = OpamFilename.create dir base in
+          let repos_roots = OpamRepositoryState.get_root st.switch_repos in
+          let src =
+            match OpamFile.OPAM.get_metadata_dir ~repos_roots opam with
+            | None ->
+              let base = OpamFilename.Base.of_string (OpamHash.contents hash) in
+              OpamFilename.create extra_files_dir base
+            | Some mdir ->
+              let files_dir = OpamFilename.Op.(mdir / "files") in
+              OpamFilename.create files_dir base
+          in
+          if OpamHash.check_file (OpamFilename.to_string src) hash then
+            OpamFilename.copy ~src ~dst
```

Comment:
I don't really understand what you try to do here. From what I see, you create the hash-named extra-file you want to copy (in the internal `extra-files` directory or directly in the source or internal repo), then copy it.
In this source preparation function, shouldn't be: 
* if the extra-file exists in the `extra-file` internal, take it from there
* otherwise take it from the `metadata_dir` (you can use `OpamFile.OPAM.get_extra_files` to not duplicate code.

Reply:
in 9d5d9d34089439e066abbd97618b9771f7a69c69 I revise the implementation. previously, opam used OpamFile.OPAM.get_extra_files, but this iterates only over the existing files in the `files` subdirectory. The new implementation iterates over the `extra_files` field in the opam file (and not over the existing files), and checks whether a file with this hash is in `extra_files` subdirectory, or if `files/<filename>` is present.

Does the revised implementation make more sense to you?